Switch to the MatrixRTCSession layer in js-sdk

dbkr opened the pull request:
Requires https://github.com/matrix-org/matrix-js-sdk/pull/3663
Fixes https://github.com/vector-im/element-call/issues/1301

dbkr committed:
WIP refactor for removing m.call events

dbkr committed:
Merge remote-tracking branch 'origin/livekit' into dbkr/matrixrtcsession

dbkr committed:
More hacking on rtcsession

dbkr committed:
Use the right sfu config hook

dbkr committed:
Use js-sdk branch

dbkr committed:
Use latest js-sdk branch

dbkr committed:
i18n

dbkr committed:
Removed now unused useGroupCall & OpenIDLoader

dbkr committed:
Remove matrix-widget-api as a dep & remove livekit stuff from widget init

We'll always have matrix-widget-api as a dep through js-sdk so also
specifyin it ourselves just means we'll end up using a different version
when the js-sdk upgrade their copy and get wierd errors. We could add a
peerDependency if we really felt the need?

dbkr committed:
i18n

dbkr committed:
Bump js-sdk

dbkr committed:
Log ghosts

and remove unsused code

dbkr committed:
Use Matrix room membership to look up room member

Rather than the matrixRTC memberships. We're essentially trusting
LiveKit's view of weho is connected here, so we may as well include
the real names of anyone we don't think is a matrixRTC participant,
for whatever reason.

dbkr committed:
Merge remote-tracking branch 'origin/livekit' into dbkr/matrixrtcsession

SimonBrandner committed:
Update js-sdk

Signed-off-by: Šimon Brandner <simon.bra.ag@gmail.com>

dbkr committed:
Manually disconnect & reconnect the livekit call if our focus changes

Without breaking the 'disconnected' screen

dbkr committed:
Update js-sdk

dbkr committed:
Merge remote-tracking branch 'origin/livekit' into dbkr/matrixrtcsession

dbkr committed:
Merge remote-tracking branch 'origin/livekit' into dbkr/matrixrtcsession

SimonBrandner commented on src/livekit/openIDSFU.ts:
```diff
@@ -14,82 +14,78 @@ See the License for the specific language governing permissions and
 }
 
+export function sfuConfigEquals(a?: SFUConfig, b?: SFUConfig): boolean {
+  if (a === undefined && b === undefined) return true;
+  if (a === undefined || b === undefined) return false;
+
+  return a.jwt === b.jwt && a.url === b.url;
```
```suggestion
  return a?.jwt === b?.jwt && a?.url === b?.url;
```

Why not ^?

dbkr replied:
I find that much harder to understand, personally, and less obvious that it's doing the thing it intends to do. I'm fine with it, but I think it could use a comment, whereas I find the previous code a bit more self-documenting.

SimonBrandner replied:
I think that for me it's quicker to read my version as "are they equal" but I don't feel strongly about it. FWIW this would also work:

```suggestion
  if (a.url !== b.url) return false;
  if (a.jwt !== b.jwt) return false

  return true;
```

dbkr replied:
Wouldn't that also need conditionals to avoid dereferencing `undefined`?

SimonBrandner replied:
I am not sure what you mean?

dbkr replied:
The first thing it does is read `a.url` but `a` could be undefined in which case that will throw an exception.

SimonBrandner replied:
Ahm rug
```suggestion
  if (a?.url !== b?.url) return false;
  if (a?.jwt !== b?.jwt) return false

  return true;
```

Ah, right!

dbkr replied:
OOB we've decided this doesn't really matter - resolving

SimonBrandner commented on src/livekit/openIDSFU.ts:
```diff
@@ -14,82 +14,78 @@ See the License for the specific language governing permissions and
+
+  const activeFocus = useActiveFocus(rtcSession);
+
+  useEffect(() => {
+    (async () => {
+      const sfuConfig = activeFocus
+        ? await getSFUConfigWithOpenID(client, activeFocus)
+        : undefined;
+      setSFUConfig(sfuConfig);
+    })();
+  }, [client, activeFocus]);
```
```suggestion
  useEffect(async () => {
    const sfuConfig = activeFocus
      ? await getSFUConfigWithOpenID(client, activeFocus)
      : undefined;
    setSFUConfig(sfuConfig);
  }, [client, activeFocus]);
```

Wouldn't this be enough?

dbkr replied:
React won't let you have an async function as an effect callback.

SimonBrandner commented on src/room/InCallView.tsx:
```diff
@@ -442,53 +424,63 @@ export function InCallView({
+  if (!id) return undefined;
+
+  const parts = id.split(":");
+  if (parts.length < 2) {
```
Should this be `3`, if the ID looks like `@simonb:element.io:FDU546DFE`?

dbkr replied:
I don't think so? There should be at least two parts: the user ID and the device ID. If there are fewer than that, it's an error.

SimonBrandner replied:
Well `@simonb:element.io.split(":") === 2`, doesn't it?

dbkr replied:
Oh, I see what you mean. Well, this was not assuming anything in particular about the format of the user IDs other than that they might contain colons, so we just take the last part off the end. Really we ought to pick a separator that doesn't appear in either, escaping/encoding the parts if necessary.

SimonBrandner replied:
I agree, we should use something different and that the code that joins the two user ID parts works but this check doesn't...

dbkr replied:
Fair enough, I guess we specifically know it's a user ID anyway so we should treat it as such.

SimonBrandner reviewed (commented):
This is looking really good, afaict

Requesting team review again as I feel like this should get more reviewers

dbkr committed:
Don't treat empty object as a valid sfu config

This was causing an extra reconnect cycle when the call was first
joined because it thought the previous SFU config was valid. This was
probably causing some client to fail to connect at all.

dbkr committed:
Merge remote-tracking branch 'origin/livekit' into dbkr/matrixrtcsession

dbkr committed:
Remove unused memberships prop

dbkr committed:
Merge remote-tracking branch 'origin/livekit' into dbkr/matrixrtcsession

dbkr committed:
yarn

dbkr committed:
Update js-sdk

dbkr committed:
Merge remote-tracking branch 'origin/livekit' into dbkr/matrixrtcsession

dbkr committed:
Merge remote-tracking branch 'origin/livekit' into dbkr/matrixrtcsession

dbkr committed:
Include the colon in the user ID in the numebr of parts we check for

dbkr committed:
Merge remote-tracking branch 'origin/livekit' into dbkr/matrixrtcsession

dbkr committed:
Merge remote-tracking branch 'origin/livekit' into dbkr/matrixrtcsession

dbkr committed:
Merge remote-tracking branch 'origin/livekit' into dbkr/matrixrtcsession

dbkr committed:
Use js-sdk from latest develop
